### DISCUSSION TITLE
How can I use external RAM for the STM32F429I-DISC1 LTDC driver

### DISCUSSION THREAD

**Discussion by jraats** [zephyrproject-rtos/zephyr#57131] [Q&A] [answered]
I am working with the STM32F429I-DISC1 development board and use the display via the LTDC driver and ili9341 driver however when I use external ram I get a linker error that the data doesn't fit.
App overlay:
```
&ltdc {
    status = "okay";
    /delete-property/ ext-sdram;
    width = < 240 >;
    height = < 320 >;

};
```
is working find as it will skip the external SDram, but my RAM usage is then 90.59%. When I add the external ram property I get the following linker error
App overlay:
```
&ltdc {
    status = "okay";
    width = < 240 >;
    height = < 320 >;

};
```
With the linker error:
```
Parsing /root/zephyr/v3.3.0/zephyr/Kconfig
Loaded configuration '/root/git/project/build/zephyr/.config'
No change to configuration in '/root/git/project/build/zephyr/.config'
No change to Kconfig header in '/root/git/project/build/zephyr/include/generated/autoconf.h'
-- Configuring done
-- Generating done
-- Build files have been written to: /root/git/project/build
[116/124] Linking C executable zephyr/zephyr_pre0.elf
FAILED: zephyr/zephyr_pre0.elf zephyr/zephyr_pre0.map /root/git/project/build/zephyr/zephyr_pre0.map 
: && ccache /root/zephyr-sdk-0.16.0/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc  -gdwarf-4 zephyr/CMakeFiles/zephyr_pre0.dir/misc/empty_file.c.obj -o zephyr/zephyr_pre0.elf  zephyr/CMakeFiles/offsets.dir/./arch/arm/core/offsets/offsets.c.obj  -fuse-ld=bfd  -Wl,-T  zephyr/linker_zephyr_pre0.cmd  -Wl,-Map=/root/git/project/build/zephyr/zephyr_pre0.map  -Wl,--whole-archive  app/libapp.a  zephyr/libzephyr.a  zephyr/arch/common/libarch__common.a  zephyr/arch/arch/arm/core/aarch32/libarch__arm__core__aarch32.a  zephyr/arch/arch/arm/core/aarch32/cortex_m/libarch__arm__core__aarch32__cortex_m.a  zephyr/arch/arch/arm/core/aarch32/mpu/libarch__arm__core__aarch32__mpu.a  zephyr/lib/libc/minimal/liblib__libc__minimal.a  zephyr/soc/arm/common/cortex_m/libsoc__arm__common__cortex_m.a  zephyr/drivers/interrupt_controller/libdrivers__interrupt_controller.a  zephyr/drivers/clock_control/libdrivers__clock_control.a  zephyr/drivers/console/libdrivers__console.a  zephyr/drivers/display/libdrivers__display.a  zephyr/drivers/gpio/libdrivers__gpio.a  zephyr/drivers/spi/libdrivers__spi.a  zephyr/drivers/serial/libdrivers__serial.a  zephyr/drivers/timer/libdrivers__timer.a  zephyr/drivers/pinctrl/libdrivers__pinctrl.a  zephyr/drivers/reset/libdrivers__reset.a  modules/stm32/stm32cube/lib..__modules__hal__stm32__stm32cube.a  -Wl,--no-whole-archive  zephyr/kernel/libkernel.a  -L"/root/zephyr-sdk-0.16.0/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.2.0/thumb/v7e-m/nofp"  -L/root/git/project/build/zephyr  -lgcc  zephyr/arch/common/libisr_tables.a  -no-pie  -mcpu=cortex-m4  -mthumb  -mabi=aapcs  -mfp16-format=ieee  -Wl,--gc-sections  -Wl,--build-id=none  -Wl,--sort-common=descending  -Wl,--sort-section=alignment  -Wl,-u,_OffsetAbsSyms  -Wl,-u,_ConfigAbsSyms  -nostdlib  -static  -Wl,-X  -Wl,-N  -Wl,--orphan-handling=warn && cd /root/git/project/build/zephyr && /usr/bin/cmake -E echo
/root/zephyr-sdk-0.16.0/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.2.0/../../../../arm-zephyr-eabi/bin/ld.bfd: warning: orphan section `.stm32_sdram2' from `zephyr/drivers/display/libdrivers__display.a(display_stm32_ltdc.c.obj)' being placed in section `.stm32_sdram2'
/root/zephyr-sdk-0.16.0/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.2.0/../../../../arm-zephyr-eabi/bin/ld.bfd: zephyr/zephyr_pre0.elf section `.stm32_sdram2' will not fit in region `IDT_LIST'
/root/zephyr-sdk-0.16.0/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.2.0/../../../../arm-zephyr-eabi/bin/ld.bfd: section .stm32_sdram2 VMA wraps around address space
/root/zephyr-sdk-0.16.0/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.2.0/../../../../arm-zephyr-eabi/bin/ld.bfd: region `IDT_LIST' overflowed by 151681 bytes
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /usr/bin/cmake --build /root/git/project/build
make: *** [common/mk/common.mk:21: debug] Error 1
```

I am not very experienced with Zephyr, so I don't know what additional information is required, but I don't understand what might be wrong. Zephyr version v3.3.0.

Thanks
Joost

**Comment by erwango** [comment, marked as answer]
Can you provide the steps to reproduce this issue ?

Using `west build -b stm32f429i_disc1 samples/drivers/display` on `zephyr-v3.3.0-2396-g99fee43f07` with `zephyr-sdk-0.15.2`, there is no issue on my side.

Note: This is also working using `zephyr-sdk-0.16.0`

Also, I'm not sure why using an overlay as default board configuration is configured to use ext-sdram.

**Comment by jraats** [comment]
Hi @erwango 
I've tried your command to test the sample driver display. And it is working. See the memory map below:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:       37188 B         2 MB      1.77%
             RAM:       23232 B       192 KB     11.82%
             CCM:          0 GB        64 KB      0.00%
          SDRAM2:        150 KB         8 MB      1.83%
        IDT_LIST:          0 GB         2 KB      0.00%
```

I looked into the sample code and saw it is using the dummy_dc driver. Does this mean that it is not using LTDC? I don't really understand why there is 150 KB allocated. The sample code doesn't specify anywhere external RAM and the dummy driver is empty. When I increase the buffer width/height in dummy_dc.overlay the RAM will also not increase (should the internal buffer in https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/display/display_stm32_ltdc.c#L403 not increase)?

When I copy the same prj.conf settings and the same overlay into my current project the error is gone. I see also an image on the display, but my RAM usage is not exactly what it should be:

```
Memory region         Used Size  Region Size  %age Used
           FLASH:      342724 B         2 MB     16.34%
             RAM:      176896 B       192 KB     89.97%
             CCM:          0 GB        64 KB      0.00%
          SDRAM2:          0 GB         8 MB      0.00%
        IDT_LIST:          0 GB         2 KB      0.00%
```

Do I miss settings/configuration which is working for all sample codes?

I'm still debugging/checking different configurations, because the error occurred when I use my old prj.conf configuration.

Thanks in advanced for the support.

edit:
Apparently I didn't removed the app.overlay. So I've added the dummy_dc.overlay and removed app.overlay. Now it is working:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:      342724 B         2 MB     16.34%
             RAM:       23232 B       192 KB     11.82%
             CCM:          0 GB        64 KB      0.00%
          SDRAM2:        150 KB         8 MB      1.83%
        IDT_LIST:          0 GB         2 KB      0.00%
```

I still not know what I'm doing. And I'm still reading documentation and 'testing' stuff. How is LTDC enabled at the moment without app.overlay. And who is configuring it now to use external RAM? Who allocates 150KB and where is that defined?

**Comment by jraats** [comment]
So I've solved the mystery. It wasn't related to extram, so I don't know why it did work with "delete-property", but the problem was that I've put some comments in the prj.conf file. These comments where generated via: "west build -t menuconfig". So when I used the example (with the config from the sample) it works, but when I used my version (with comments) it didn't work. I've changed everything back without the comments in the prj.conf file.

Also I've overwritten the LTDC driver (via app.overlay) and changed the width and height. I see that the SDRAM2 also increases now (more width/height is also more bytes required (w*h*2). I had changed in my previous question the width and height of the dummy_dc driver and I think that one is disabled, because the stm32 dev kit contains an LTDC chip? It is for me still unknown when which driver is enabled. But my initial problem is solved now.

I'm going to read more info about zephyr overlay/device tree, because some parts are still confusing.

Thanks for the help @erwango 

Joost
